docs/rfc/012-record-transaction-model.md

```diff
@@ -0,0 +1,243 @@
+a Leo program must include a function called `main`, in its aforementioned main file.
+The processing of a transaction corresponds to an invocation of this `main` function.
+The word 'corresponds' in the preceding sentence is important:
+unlike other blockchains like Ethereum,
+the processing of the transaction does not involve executing the Leo code;
+rather, it involves checking a zero-knowledge proof
+of the execution of the Leo program,
+which was prepared when the Leo program was compiled.
+This is what 'corresponds' means, in that sentence.
+However, for the high-level purpose of this RFC, these are zero-knowledge details.
+
+In general, the `main` function takes some `const` and some non-`const` inputs (declared as parameters),
+and returns an output (declared as a return type), which may be a tuple to effectively represent multiple outputs.
+The `const` inputs are compiled into the zero-knowledge circuit,
+so they can be ignored for our purpose here,
+leaving the non-`const` inputs and the output for consideration.
+
+The execution of `main` can be described as a mathematical function
+```
+main : Record x Record x Inputs -> Record x Record x Output
+```
+where `x` is cartesian product,
+`Record` is the set of possible records,
+`Inputs` is the set of possible inputs to `main`, and
+`Output` is the set of possible outputs from `main`.
+(These sets can be thought as "types", but mathematically we talk about sets.)
+That is, this mathematical function
+takes three inputs (the two old records and the `main` inputs)
+and returns three outputs (the two new records and the `main` output).
+While `Record` is fixed, i.e. it is the same for all Leo programs,
+both `Inputs` and `Output` differ based on the Leo input and output types of `main`.
+
+In the Leo code, in `main` or in functions called by `main`,
+the values in `Inputs` are accessed via the `main` parameters,
+while the old records are accessed via the special `input` variable,
+which provides access to the two old records and their slots,
+including the payloads that contain application-specific data.
+The picture for new records and values in `Output` is less clear from the documentation:
+experimentation suggests that the new records are obtained
+by serializing the output value in `Output` (which, recall, may be a tuple).
+
+It is important to note that the values in `Inputs` do not come from the two old records.
+Rather, they are private inputs, supplied by the developer
+when they compile the Leo program and generate the zero-knowledge proof.
+Indeed, as mentioned above, the processing of the transaction in the blockchain
+does not execute the Leo code, and thus does not need to know the values in `Inputs`.
+Rather, the blockchain has to verify a zero-knowledge proof asserting that
+there exist values in `Input`, known to the creator of the transaction,
+such that the execution of the Leo program's `main`
+on those values and on the old records
+yields the new records;
```

Comment:
I pushed a commit that mentions 'some output'. I didn't initially mention it because (my understanding is that) only the records matter for the zk proof, as only records are represented in the blockchain, and so the zk assertion must talk about old and new records. That's why the inputs are existentially quantified, and so is the output, but the output is calculated from the inputs, so there's no need to existentially quantify it. However, based on your comment, it's best to explicate it, which the new commit does.